src/Prefix.jl

```diff
@@ -228,19 +228,64 @@ function extract_platform_key(path::AbstractString)
     return platform_key(path[idx+1:end])
 end
 
+"""
+    isinstalled(tarball_url::AbstractString,
+                hash::AbstractString;
+                prefix::Prefix = global_prefix)
+
+Given a `prefix`, a `tarball_url` and a `hash`, check whether the
+tarball with that hash has been installed into `prefix`.
+
+In particular, it checks for the tarball, matching hash file, and manifest
+installed by `install`, and checks that the files listed in the manifest
+are installed and are not older than the tarball.
+"""
+function isinstalled(tarball_url::AbstractString, hash::AbstractString;
+                     prefix::Prefix = global_prefix)
+    # check that the hash file and tarball exist and match hash
+    tarball_path = joinpath(prefix, "downloads", basename(tarball_url))
+    hash_path = "$(tarball_path).sha256"
+    if safe_isfile(tarball_url)
+        tarball_path = tarball_url
+    end
+    isfile(tarball_path) || return false
+    isfile(hash_path) || return false
+    read(hash_path, String) == hash || return false
+    tarball_time = stat(tarball_path).mtime
+    stat(hash_path).mtime >= tarball_time || return false
+    # verify(...) doesn't re-hash if the hashfile matches, so we won't either
+    # bytes2hex(open(sha256, tarball_path)) == hash || return false
```

Comment:
Why not just use `verify(tarball_path, hash)` here?  We don't really care if the `hash_path` is missing (it shouldn't be, but if it is, that doesn't necessarily mean that the queried tarball is not installed); we can still verify the tarball and come to a supposedly valid conclusion.

Reply:
I initially did that, but I noticed that `verify` has two problems:

1. It doesn't return `false` if the `.sha256` doesn't exist or fails to match `hash` — it just goes ahead and hashes the tarball.   That's because `verify` only cares that the tarball is correct.  `isinstalled`, on the other hand, is trying to check whether `install` was run successfully with this `hash`, in which case it I think we *do* case if the `hash_path` is missing/wrong because a successful `install` should have always created the `.sha256` file.

2. It overwrites the `.sha256` at the end, whereas I felt that `isinstalled` should be read-only.

There's not much duplication here, anyway — it's only a couple of lines in `isinstalled`.

Reply:
I think I would prefer `verify()` to be used here.  We don't really care about the state of the `.sha256` files, we care about the state of the `.tar.gz` file.  If the user did something strange like delete a `.sha256` file but not the tarball, I think it makes sense for `isinstalled()` to recreate the `.sha256` file.  It's just a performance enhancement to stop us from having to rehash the `.tar.gz` file every time we look at it; its presence or staleness should have no semantic value, and should be transparently updated as often as needed.

Trying to use the presence of a `.sha256` file to determine whether a tarball has been installed or not isn't perfect either.  Really, we're missing the step where we extract the files from a tarball, hash them, and then compare hashes with the files that exist in the prefix.  But that will be way too slow and too cumbersome.  Alternatively, we could embed the tarball hash into the manifest that we write out after installation; then we could skip checking the tarball hash at all, because that's not really what we care about; it's the state of the files that were extracted that we care about, and the manifest is a much more direct representation of that than the state of the tarball, although I do admit that it's not that much of an improvement.

Reply:
I was about to make this change, but I noticed a third problem, this one insurmountable: `verify` throws an error if the `.tar.gz` doesn't match the hash, whereas `isinstalled` should just return `false`.